gensim/models/nmf.py

```diff
@@ -458,90 +561,88 @@ def _setup(self, corpus):
             )
         )
 
-        is_great_enough = self._W > self.w_std * self.sparse_coef
+        self.A = np.zeros((self.num_topics, self.num_topics))
+        self.B = np.zeros((self.num_tokens, self.num_topics))
 
-        self._W *= is_great_enough | ~is_great_enough.all(axis=0)
-
-        self._W = scipy.sparse.csc_matrix(self._W)
-
-        self.A = scipy.sparse.csr_matrix((self.num_topics, self.num_topics))
-        self.B = scipy.sparse.csc_matrix((self.num_tokens, self.num_topics))
-
-    def update(self, corpus, chunks_as_numpy=False):
+    def update(self, corpus):
         """Train the model with new documents.
 
         Parameters
         ----------
-        corpus : iterable of list(int, float)
+        corpus : iterable of list of (int, float), optional
             Training corpus.
-        chunks_as_numpy : bool, optional
-            Whether each chunk passed to the inference step should be a numpy.ndarray or not. Numpy can in some settings
-            turn the term IDs into floats, these will be converted back into integers in inference, which incurs a
-            performance hit. For distributed computing it may be desirable to keep the chunks as `numpy.ndarray`.
+            Can be either iterable of documents, which are lists of `(word_id, word_count)`,
+            or a sparse csc matrix of BOWs for each document.
+            If not specified, the model is left uninitialized (presumably, to be trained later with `self.train()`).
 
         """
-
         if self._W is None:
             self._setup(corpus)
 
         chunk_idx = 1
 
         for _ in range(self.passes):
-            for chunk in utils.grouper(
-                corpus, self.chunksize, as_numpy=chunks_as_numpy
-            ):
-                self.random_state.shuffle(chunk)
-                v = matutils.corpus2csc(chunk, len(self.id2word)).tocsr()
-                self._h, self._r = self._solveproj(
-                    v, self._W, r=self._r, h=self._h, v_max=self.v_max
+            if isinstance(corpus, scipy.sparse.csc.csc_matrix):
+                grouper = (
+                    corpus[:, col_idx:col_idx + self.chunksize]
```

Comment:
This may be due to some changes in scipy, but your code results in

```
IndexError: index out of bounds: 0 <= 5 <= 10, 0 <= 100500 <= 10, 5 <= 100500
```

in scipy 0.19.0.

Reply:
Oh, I see. Seems like they added the support for that in the following versions, I'll do something like `corpus[:, col_idx:min(col_idx + self.chunksize, chunk.shape[1])]`

Reply:
Thanks. Plus add a comment, so someone doesn't accidentally remove it again in the future. 

We may simplify it again once the older scipy versions become irrelevant in the future.